src/Logic.lidr

```diff
@@ -0,0 +1,1296 @@
+
+
+==== Exercise: 1 star, optional (iff_properties)
+
+Using the above proof that \idr{<->} is symmetric (\idr{iff_sym}) as a guide,
+prove that it is also reflexive and transitive.
+
+>   iff_refl : p <-> p
+>   iff_refl = ?iff_refl_rhs
+
+>   iff_trans : (p <-> q) -> (q <-> r) -> (p <-> r)
+>   iff_trans piq qir = ?iff_trans_rhs
+
+$\square$
+
+
+==== Exercise: 3 stars (or_distributes_over_and)
+
+>   or_distributes_over_and : (p `Either` (q,r)) <-> (p `Either` q, p `Either` r)
+>   or_distributes_over_and = ?or_distributes_over_and_rhs
+
+$\square$
+
+\todo[inline]{Edit the rest of the section. What to do with Setoids? We could
+probably just use profunctors here}
+
+Some of Idris's tactics treat iff statements specially, avoiding the need for
+some low-level proof-state manipulation. In particular, rewrite and reflexivity
+can be used with iff statements, not just equalities. To enable this behavior,
+we need to import a special Idris library that allows rewriting with other
+formulas besides equality:
+
+Require Import Idris.Setoids.Setoid.
+
+Here is a simple example demonstrating how these tactics work with iff. First,
+let's prove a couple of basic iff equivalences...
+
+>   mult_0 : (n * m = Z) <-> ((n = Z) `Either` (m = Z))
+>   mult_0 {n} {m} = (to n m, or_example n m)
+>   where
+>     to : (n, m : Nat) -> (n * m = Z) -> (n = 0) `Either` (m = 0)
+>     to Z _ Refl = Left Refl
+>     to (S _) Z _ = Right Refl
+>     to (S _) (S _) Refl impossible
+
+>   or_assoc : (p `Either` (q `Either` r)) <-> ((p `Either` q) `Either` r)
+>   or_assoc = (to, fro)
+>   where
+>     to : Either p (Either q r) -> Either (Either p q) r
+>     to (Left p) = Left $ Left p
+>     to (Right (Left q)) = Left $ Right q
+>     to (Right (Right r)) = Right r
+>     fro : Either (Either p q) r -> Either p (Either q r)
+>     fro (Left (Left p)) = Left p
+>     fro (Left (Right q)) = Right $ Left q
+>     fro (Right r) = Right $ Right r
+
+We can now use these facts with rewrite and reflexivity to give smooth proofs of
+statements involving equivalences. Here is a ternary version of the previous
+mult_0 result:
+
+>   mult_0_3 : (n * m * p = Z) <-> 
+>              ((n = Z) `Either` ((m = Z) `Either` (p = Z)))
+>   mult_0_3 = (to, fro)
+>   where 
+>     to : (n * m * p = Z) -> ((n = Z) `Either` ((m = Z) `Either` (p = Z)))
+>     to {n} {m} {p} prf = let 
+>       (nm_p_to, _) = mult_0 {n=(n*m)} {m=p}
+>       (n_m_to, _) = mult_0 {n} {m}
+>       (_, or_a_fro) = or_assoc {p=(n=Z)} {q=(m=Z)} {r=(p=Z)}
+>       in or_a_fro $ case nm_p_to prf of 
+>            Left prf => Left $ n_m_to prf
+>            Right prf => Right prf
+>     fro : ((n = Z) `Either` ((m = Z) `Either` (p = Z))) -> (n * m * p = Z)
+>     fro (Left Refl) = Refl
+>     fro {n} (Right (Left Refl)) = rewrite multZeroRightZero n in Refl
+>     fro {n} {m} (Right (Right Refl)) = rewrite multZeroRightZero (n*m) in Refl
+
+The apply tactic can also be used with <->. When given an equivalence as its
```

Comment:
Yeah, I didn't really format these paragraphs because they'll have to be rewritten or removed anyways - they speak about how awesome is setoid manipulation with tactics.